datalake/src/main/java/com/sequenceiq/datalake/service/recovery/SdxRecoverySelectorService.java

```diff
@@ -0,0 +1,69 @@
+package com.sequenceiq.datalake.service.recovery;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.dto.NameOrCrn;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.sdx.api.model.SdxRecoverableResponse;
+import com.sequenceiq.sdx.api.model.SdxRecoveryRequest;
+import com.sequenceiq.sdx.api.model.SdxRecoveryResponse;
+
+@Service
+/**
+ * Recovers an SDX cluster from a failure during Resize or Upgrade.
+ *
+ * Chooses the appropriate recovery action to take based on the state of the SDX cluster. Examples:
+ * <ul>
+ *     <li>{@Code SdxUpgradeRecoveryService}</li>
+ *     <li>{@Code ResizeRecoveryService}</li>
+ * </ul>
+ */
+public class SdxRecoverySelectorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SdxRecoverySelectorService.class);
+
+    private List<RecoveryService> services;
+
+    @Autowired
+    public SdxRecoverySelectorService(List<RecoveryService> services) {
+        this.services = services;
+    }
+
+    public SdxRecoveryResponse triggerRecovery(String crn, NameOrCrn nameOrCrn, SdxRecoveryRequest sdxRecoveryRequest) {
+
+        List<String> responseReasons = new ArrayList<>();
+
+        for (RecoveryService recoveryService : services) {
+            SdxRecoverableResponse recoverableResponse = recoveryService.validateRecovery(crn, nameOrCrn);
+            responseReasons.add(recoverableResponse.getReason());
+            if (recoverableResponse.getStatus().recoverable()) {
+                return recoveryService.triggerRecovery(crn, nameOrCrn, sdxRecoveryRequest);
+            }
+        }
+
+        String reasons = String.join(",", responseReasons);
+        LOGGER.debug("Cluster is not in a recoverable state with message(s): {}", reasons);
+        throw new BadRequestException(reasons);
+
+    }
+
+    public SdxRecoverableResponse validateRecovery(String crn, NameOrCrn nameOrCrn) {
+
+        SdxRecoverableResponse sdxRecoverableResponse = null;
+        for (RecoveryService recoveryService : services) {
+            if (recoveryService.canRecover(crn, nameOrCrn)) {
+                sdxRecoverableResponse = recoveryService.validateRecovery(crn, nameOrCrn);
+                if (sdxRecoverableResponse.getStatus().recoverable()) {
+                    return sdxRecoverableResponse;
+                }
+            }
+        }
+        return sdxRecoverableResponse;
```

Comment:
This will return a null response if there are no services that can recover the resource, is this okay?

Reply:
Ah whoops my mistake, it will just keep the last service's response which indicates that it is not recoverable.

I guess instead let me ask: Is it okay to just return one service's response? Or would it be better to build a response indicating that recovery is not possible for any of the available services?

I ask this since a user that tries to recover a failed resize for instance may be very confused if they receive a message saying that Upgrade can not be recovered.

Reply:
added a case to cover null here, because it actually might be possible. Since we are checking the last failed chain in canRecover() for now we only should have one recovery service trying to respond. In the future it might be valid that we have more than one way to recover, but we'd also need to expand the API to reflect that.

Reply:
Sounds good thanks!